### DISCUSSION TITLE
Possible to update MQTT based on value changes?

### DISCUSSION THREAD

**Discussion by AleXSR700** [arendst/Tasmota#16458] [Q&A]
Hello everyone,
I am using multiple tasmota devices to monitor power consumption as well as relay temperature and humidity information and TRV information.

Right now I have to manually set TelePeriod and Mi32Period. This causes either a lack of data for quickly changing devices (e.g. microwaves) or an excess of data for slowly changing devices (e.g. microwave when not in use).
So either I set a short teleperiod (e.g. 10 seconds) and am filling my database with useless information like standby power consumption which does not change OR I am missing the data I need because I set teleperiod to e.g. standard value of 300 s when of course the microwave only runs for ~60-120 seconds.

Most devices like Shelly etc. dedicated to power consumption monitoring always broadcast new data when the value changes. Would something like this be possible with tasmota?
Possible ideas:
- Broadcast new value when value changes by 1 unit
or
- Broadcast new value when value changes by user defined amount

Ideally the reference value can/must be selected, i.e. only broadcast if value of temperature or humidity changes but not when useless information like time or certain states change.

Any ideas would be appreciated.

Thank you all and best regards
Alex

**Comment by sfromis** [comment]
When it comes to energy monitoring, you have `PowerDelta` to trigger reporting when consumption level change, working well with a longer TelePeriod. As a starting point, I'd suggest something like `PowerDelta 110` for a 10 watt change as a trigger, obviously depending on the consumer.

**Reply by AleXSR700** [reply]
Thank you @sfromis , that looks excellent! I will try it over the next few days :)
Is there such a feature for non-power measurements like humidity or temperature? I did not find one in 'commands'.

EDIT:
I just found that PowerDelta is not supported by ESP32 devices (I am using ESP32 bluetooth firmware). Is this because of the bluetooth version?
It seems to work on my tasmota sockets though. So let's see how that behaves. Would be great to have this also for temperature & humidity.

**Reply by sfromis** [reply]
PowerDelta is very much supported on ESP32, including with Bluetooth (just verified with `tasmota32-bluetooth.bin`), but in general only when Tasmota is handling the energy monitoring locally on the device.

For other sensors there is no feature similar to PowerDelta, but with ESP32, the Berry language makes such things feasible to create without a lot of code.

**Reply by AleXSR700** [reply]
Ah, thank you very much. Indeed the ESP32 only acts as a hub and does not measure anything itself. 

I will start reading up on the Berry scripting language on the Tasmota website. Not sure I will figure out this concrete problem, but I will try.

**Reply by barbudor** [reply]
You can have sensors on ESP32 too
On any esp you fan also use Rules and there are exemples in the cookbook to send sensor data when change is over a given threshold

**Reply by AleXSR700** [reply]
Yes, thank you, I saw those. But is there a "friendly_name" or only the device name?
Example:

```18:28:09.539 MQT: tele/tasmota_ble/ATC76598c = {"Time":"2022-09-08T18:28:09","mac":"a4c13876598c","Temperature":21.8,"Humidity":54.1,"DewPoint":12.1,"Btn":1,"Battery":81,"RSSI":-73,"TempUnit":"C"}```

So would it be "ATC76598C_Temperature" for the sensor I want to check?

**Reply by sfromis** [reply]
To get a "friendly name" use `BLEAlias` to assign names.

**Reply by barbudor** [reply]
You won't be able to get rules to work with your MI32Options. You need to have the device ID as a Key in the JSON so you can use a trigger such as `ON deviceid#temperature DO something ... ENDON`
Exemple:
```
Rule1 on ATCc09ad3#temperature do var9 %value% endon

19:25:19.475 MQT: tele/devble/chbre_manon = {"Time":"2022-09-08T19:25:19+02:00","ATCc09ad3":{"alias":"chbre_manon","mac":"a4c138c09ad3","Temperature":25.9,"Humidity":53.2,"DewPoint":15.6,"Btn":1,"Battery":80,"RSSI":-75},"TempUnit":"C"} (retained)
19:25:19.510 RUL: ATCC09AD3#TEMPERATURE performs "var9 25.9"
19:25:19.519 MQT: stat/esp32dev3/VAR = {"Var9":"25.9"}
```
To get the above, I'm using:
```
Mi32Option0 1
Mi32Option1 0
Mi32Option2 0
Mi32Option4 1
Mi32Option5 1
Mi32Option6 2
```
Oups, it looks like I forgot to update the docs regarding `Mi32Option6 2` : https://github.com/arendst/Tasmota/pull/15383

Alternatively as you are on ESP32, using Berry may be more usefull as you can keep your existing configuration and process the full JSON.

However, I thought that you were looking for a solution working with native sensors.
The kind of rule described [here](https://tasmota.github.io/docs/Rules/#transmit-sensor-value-only-when-a-delta-is-reached) is usefull only with native sensors because the Rules engine is called continuously, depending on sensors it could be more than once per second. So you can catch changes and publish on change.

With BLE sensors it is different because tasmota receive data when the sensor is sending it and this is not under the control of Tasmota. Most of the time BLE sensors are sending datas every minutes or so otherwise their batteries would drain quickly.
And the Rules engine is called only at the same time the data is published on MQTT.
So you can't do the same with BLE sensors as with native sensors

The best you can do with BLE sensors is to use `Mi32Option2 1` to force Tasmota to ignore Mi32Period and send sensor data as soon as it is received.

**Reply by AleXSR700** [reply]
I am using only BLE sensors.
But the BLE sensors send every 10 seconds and with Mi32Period I can set how often this data is being forwarded.

But instead of using a short Mi32Period, I would like to set Mi32Period to 10min and in addition have a Berry Script send additional packages whenever the temperature or humidity changes by a defined value.
And that is where I am struggling right now.

EDIT & PS:
I am trying to set a long MI32Period (so only send an update every 10min) but send the full package whenever there is a change of temperature or humidity greater than x.

So I do not want to define a new Var, I want to then send the entire "normal" tele package.

**Reply by AleXSR700** [reply]
I could not find the solution in the docs examples.

How can I check value **change**?

```Rule1 on ATCc09ad3#temperature do var9 %value% endon```
This would trigger everytime the temperature changes. But how can I adjust this to only trigger when the change is larger than e.g. 2?
Or would I need to create a new variable and then compare it?

```
Rule1 
ON ATCc09ad3#temperature DO var1 %value% ENDON
ON ATCc09ad3#temperature IF (var1 - %value%)>2 DO publish tele/tasmota_ble/ATC76598c/temperature ENDIF ENDON
```

**Reply by barbudor** [reply]
I already provided you with the link on how to do that 4 days ago

**Reply by AleXSR700** [reply]
Good morning @barbudor,
Yes, I saw that but you said it was only for native sensors. I thought it would then not apply to BLE sensors.

In addition, I was hoping that there was a way that did not need the creation of variables.
I have 8 sensors right now and I would like to monitor 3 values each. So I would need to write 24 variables. Only 16 are possible, if I understand the docs correctly. This also means that I cannot add more sensors later or make any other rules requiring variables to be defined.

So am I correct that I have to find a solution using the (only 16) variables? There is no built-in "compare to previous" like the PowerDelta uses?

**Reply by sfromis** [reply]
No built-in "compare for change". As long as you can have a trigger like `ON ATCc09ad3#temperature DO` work, it should work just like a native sensor when it comes to rules.

A good solution, as barbudor already told you, would be to use the greatly expanded features of the Berry language instead of "classic" rules.

**Reply by AleXSR700** [reply]
I would love to use Berry, but I am completely unfamiliar with the syntax there.

For normal rules I think this should be the correct approach. But I am not sure if my checking of the variable is correct ("check if var1 is empty"), but I think an empty variable is a string, so it hopefully is correct.
If it is, this should 
1. If temperature or humidity changes
2. Check if variable has a value.
3. If variable is empty then set it to current value
4. If the variable has a value then
4.1 Check if the value has changed by more than 1 (increased or decreased)
4.2 Publish the entire sensor data
4.3 Set the current value as the new variable
5. If the value has changed by less than 1, do nothing

```
Rule 1
ON SENSOR1#temperature DO IF var1="" var1 %value% ELSEIF (%value% - var1)>1 OR (var1 - %value%)>1 publish tele/SENSOR1; var1 %value% ENDIF ENDON
ON SENSOR1#humidity DO IF var2="" var2 %value% ELSEIF (%value% - var2)>1 OR (var2 - %value%)>1 publish tele/SENSOR1; var2 %value% ENDIF ENDON
```

But what would Berry add in this case? Do I not still need to create variables and am I not also limited to 16 variables?
Upside would probably be that I do not define this for all sensors individually but could have the script do this per sensor automatically. That is if I knew how to :D

**Reply by sfromis** [reply]
It seems that you are also not too familiar wjith the syntax for "classic" rules either :wink: 

With Berry code, you have standard programming facilities, and can have hundreds of variables, including maps, meaning that you could code difference detection without creating separate variables/code for every single value. Much easier than classic Tasmota rules.

But yeah, anything takes a learning curve to get the hang of it. The Berry cookbook page does provide quite a few code examples illustrating the general concepts, and of course there is also docs for the language.

**Reply by AleXSR700** [reply]
Indeed I am new to Tasmota rules also. It seems more or less everything and everyone uses their own variation of the same code. So what am I missing in my code above?

I looked at the Cookbook before but my first thought was "oh god, not another non-intuitive language" 😄 
As a non-programmer who has to work with javascript, python, vba and others (worst of all jinja2), every new syntax that is not "similar" to my "known" ones becomes a pain. 😄 
And like Jinja2, Berry seems to have few active communities out there where a complete noob can find help.

The Cookbook, by the way, is not very noob friendly nor does it have any simple examples like creating replacements for rules 😉

**Reply by AleXSR700** [reply]
I think the variable value must be in %..% for calculations and I fogrot the parentheses.
```
​Rule 1
ON SENSOR1#temperature DO IF (var1="" var1) %value% ELSEIF ((%value% - %var1%)>1 OR (%var1% - %value%)>1) publish tele/SENSOR1; var1 %value% ENDIF ENDON
ON SENSOR1#humidity DO IF (var2="" var2) %value% ELSEIF ((%value% - %var2%)>1 OR (%var2% - %value%)>1) publish tele/SENSOR1; var2 %value% ENDIF ENDON
```

**Reply by techman83** [reply]
You'll find Berry is most like python as far as languages go. It is fairly new as far as languages go, but it is much easier that writing C/C++!

I'm sure contributions would be welcome as far as documentation goes 🙂

**Reply by sfromis** [reply]
Here's a small Berry example of handling temperature for a particular sensor:
```
tasmota.add_rule('BME280#Temperature',
  def (value,trigger,jsonmap)
    if value<20
      print('getting cool')
    else
      print('OK-ish')
    end
  end)
```
Details can vary depending on the specifics. Note that `jsonmap` includes the full JSON payload parsed.

**Reply by AleXSR700** [reply]
I will try to read up on Berry scripting and then share my progress here. It would be great if you would continue to check in and support as kindly as you have been :)

@sfromis:
Your example creates native tasmota rules, does it not? If so, will it not be subject to the exact same limitations (e.g. number of variables usable)?
So what I would need to do is write a script that does (I will try to put it in a mix of plain text and "code"
```
def TemperatureDelta()
  For i in Connected_BLE_devices Do
    If Connected_BLE_deviceNameVar ="" Then Connected_BLE_deviceNameVar = current_temperature_value
    Else
      If (Connected_BLE_deviceNameVar - current_temperature_value)>1 Or (current_temperature_value - Connected_BLE_deviceNameVar)>1 Then
        mqtt.publish fullpayload
        Connected_BLE_deviceNameVar = current_temperature_value
      End If
    End If
  End
```
So I need to find out how to get a list of sensors (maybe with tasmota.read_sensors or introspect or maybe there is something similar to tasmota.get_switches?).
Then create a variable and perform the comparison.
Then publish full payload if condition is met.

**Reply by barbudor** [reply]
You don't need to get a list of sensors in advance
You get the json, extract the name and you can store temp hi/low into a 'map' which is an array indexed not by numbers but by name. Other languages such as python call that dictionnaires
A JSON is a representation of a map. So you can see map as a variable storing a json, not in its ASCII form but as an object form
The Berry rule function receive as its 3rd parameter such a map holding the JSON on the MQTT payload that triggered the rule

You would end up with a Berry global variable for example such as 'Temp' and use it as `Temp[sensor_name]['hi']` where `sensor_name` would be extracted from the 'name' key of the BLE message

**Reply by AleXSR700** [reply]
Ah, that is good to know :)
Are you suggesting one map per device or one map for all devices?
I would like to make sure that the script is used for all current and future devices connected. So I do not want to define the devices' names but basically add every current (or new) device to the routine automatically. 

I think Tasmota might already have that kind of map of devices because all BLE devices are shown on the starting page of the WebUI. So the data is already being processed. Is this map of names maybe already available to access via Berry script? Or do I need to create that list/map myself as you mentioned from the json?


P.S.: I would like to always publish the full information of the respective BLE device, not just the criteria value. So if temperature meets the criteria and triggers the action, all other parameters shall be published as well.  They change also but are less important to me or are directly related to the criteria (e.g. dew point is a function of temperature and relative humidity, so I do not need to monitor its change but I do want to publish it also).

**Reply by sfromis** [reply]
No, you need no map of existing sensor names, each payload has info about what's included of sensors and values. Note that with BLE, you can vary the JSON format by options, including but not limited to a format matching native sensors, but as long as you code for the JSON format available, which Tasmota always parses into a map for easy access, this is no requirement.

What you may want to have is a list of which JSON keys are interesting to you, like from this sample BLE payload:
`{"Time":"2022-09-13T09:53:47.810+02:00","LYWSD038e0a7d":{"alias":"Temp1","mac":"a4c1388e0a7d","Temperature":24.70,"Humidity":54.90,"DewPoint":15.01,"Battery":73,"RSSI":-81},"TempUnit":"C"}`

**Reply by AleXSR700** [reply]
To me only 'Temperature' and 'Humidity' are relevant. If work load is negligible  for additional criteria,  I would add 'DewPoint'.


> Note that with BLE, you can vary the JSON format by options, including but not limited to a format matching native sensors, but as long as you code for the JSON format available, which Tasmota always parses into a map for easy access, this is no requirement.

This is interesting. Does this mean that I can access the Temperature, Humidity and DewPoint with a built-in command (like `sensor1#temperature`) because they are being recognized by Tasmota and processed for display on main page already?

**Reply by AleXSR700** [reply]
So, 
1. Trigger: everytime a payload is received
2. Get `Name` from jsonmap
3. Get `Temperature` from jsonmap
4. Check if `Name` has an existing variable called `Name`+'tempvar' (create one if not)
5. Compare existing value of respective tempvar with `Temperature`
6. Perform action (mqtt publish full payload)

**Reply by sfromis** [reply]
No built-in **command**. It means that the JSON format can be varied, which means that the Berry code has to match the format you've chosen to use.

With my sample payload above, you could have a trigger (not a command) like `LYWSD038e0a7d#Temperature` to match the JSON object hierarchy. This is not directly related to what's visible on the main page, but a function of what you have in the JSON being published. Watch the console instead of the main page.

**Reply by AleXSR700** [reply]
But if I want to watch the console for such a trigger, I must know the name of the sensor. So I would need to define the name first. Otherwise I would need to, as in my previous "steps" post suggested, trigger on every payload reception.
Correct?

I do not see a rule trigger for the reception of a payload from *any* BLE device https://tasmota.github.io/docs/Rules/#rule-trigger
They seem to always require information of specific `name` or specific `sensor`.

P.S.: To ensure that the json format matches the payload/source, I could filter by grouptopic. My Xiaomi Mijia devices will be one format and my eQ-3 TRVs (which are grouptopic2 eq3_devices) could have a second json format.

**Reply by barbudor** [reply]
There is not directly a trigger for _any_ BLE sensor but you can use a trick : triggers on the `Time` key and then parse the jsonmap yourself

This should do what you are looking for:
I left a lot of print so you see what happens
```python
import math
import string
BleTH = {}
tasmota.add_rule("Time",def (value,trigger,jsonmap)
  print("BleRule:", jsonmap) 
  for key:jsonmap.keys()
    if key[0..2] == "ATC" || key[0..4] == "LYWSD"
	  print(string.format("found key %d", key))
	  if !BleTH.contains(key)
	    print("  adding to BleTH")
	    BleTH[key] = { "Temperature": 0.0, "Humidity": 0.0 }
	  else
		print(string.format("  previous data {'Temperature':%f,'Humidity':%f}", BleTH[key]['Temperature'], BleTH[key]['Humidity']))
	  end
	  var temp = jsonmap[key]['Temperature']
	  var hum = jsonmap[key]['Humidity']
	  if math.abs(temp-BleTH[key]['Temperature']) > 0.5 || math.abs(hum-BleTH[key]['Humidity']) > 0.5
		BleTH[key] = { "Temperature": temp, "Humidity": hum }
		print(string.format("  %s: {'Temperature':%f,'Humidity':%f}", key, temp, hum))
	  end
	end
  end
end)
```

**Reply by AleXSR700** [reply]
Thank you so very, very much, @barbudor 😍 
👏 👏 👏 👏 

Never in a million years would I have managed to get that right myself 🙈 

I had to try it right away and try to understand it all as best I could. But there is one thing I am not sure if I am missing the automated trigger or if this is not a part of the script:
For the script to be able to read data and process it, it seems that the `Mi32Period`has to be set to a small value, i.e. unless there is a payload being sent via MQTT, the script will not process anything.
This sounds kind of logical, but the information is being received by the ESP32 from the Xiaomi devices every 10 seconds. So the information should be there. Only the forwarding (via MQTT) is not being performed (which would be good, because ideally the `Mi32Period` is set to 3600 (just like the TelePeriod) but the script processes the incoming data everytime it is received).

So does the script only see what I see in the console? The data received in the background, i.e. the sensor data that is coming in every 10 seconds is not visible to the script?
Or am I missing a `SetOption` or one of the undocumented `Mi32Option` (Mi32Options range from 1 to 6 but 3 and 4 are not documented)?
At the moment I have no `SetOption` or `Mi32Option` changed from default except `Mi32Option6 2`.

**Reply by barbudor** [reply]
The BLE driver calls the Rules engine (including Berry rules) only on MQTT publish
It doesn't call it on data received from the sensor.

You may try `Mi32Option2 1` and `Mi32Option3 1` to have the publication and the rule processing everytime a message is received
You can then publish the changes on a different topic than SENSOR and forget SENSOR, using only your change topic

**Reply by AleXSR700** [reply]
Direct bridging was something I considered, but I need to test again if that overrides Mi32Period.
The `Mi32Option2` states "sensor data only at TELEPERIOD (default and "usual" Tasmota style)" but in fact, if my observations are correct, it is not the `TELEperiod` but rather the `Mi32Period` that publishes the sensor data, so I will need to check if this setting then successfully overrides Mi32Period setting and triggers the script.

Quick question: `Mi32Option3` is undocumented (same for `Mi32Option4`). What is Mi32Option3?

**Reply by barbudor** [reply]
Yes, in Mi32option2 read 'mi32period' instead of teleperiod

Mi32Option3 does nothing. No code.

Feel free to improve the doc by clicking on the pencil button on the top right of the page

**Reply by barbudor** [reply]
Looks like I made errors in my post
You need to try MI32Option1 1 and Mi32option2 1
